src/test/kotlin/org/rust/lang/core/type/RsGenericExpressionTypeInferenceTest.kt

```diff
@@ -541,4 +541,14 @@ class RsGenericExpressionTypeInferenceTest : RsTypificationTestBase() {
           //^ (Bar<(i32, bool), i32>, Foo<[f64; 3], &str>)
         }
     """)
+
+    fun `test infer generic argument from trait bound`() = testExpr("""
+        struct S<X>(X);
+        trait Tr<Y> { fn foo(&self) -> Y; }
+        impl<Z> Tr<Z> for S<Z> { fn foo(&self) -> Z { unimplemented!() } }
+        fn bar<A, B: Tr<A>>(b: B) -> A { b.foo() }
+        fn main() {
+            bar(S(1))
+        }    //^ i32
```

Comment:
Let's change this a little bit such that it is not ambiguous which expression gets selected by `^` :

```
let x = bar(S(1)); x
                 //^ i32
```